File: d365/FileWatcherMonitoring.Plugins/StateTransitionPolicy.cs
using System.Collections.Generic;

namespace FileWatcherMonitoring.Plugins
{
    /// <summary>
    /// Port of engine/state-transition.policy.ts. Allow-list, not deny-list:
    /// any transition not listed is invalid.
    /// </summary>
    public static class StateTransitionPolicy
    {
        private static readonly Dictionary<string, FileStatus[]> ValidTransitions =
            new Dictionary<string, FileStatus[]>
            {
                { None, new[] { FileStatus.FILE_DETECTED, FileStatus.FILE_MISSING_BY_SLA } },
                { nameof(FileStatus.FILE_DETECTED), new[] { FileStatus.FILE_STABLE, FileStatus.FILE_STUCK } },
                { nameof(FileStatus.FILE_STABLE), new[] { FileStatus.FILE_DUPLICATE } },
                { nameof(FileStatus.FILE_STUCK), new[] { FileStatus.FILE_STABLE } },
                { nameof(FileStatus.FILE_DUPLICATE), new FileStatus[0] },
                { nameof(FileStatus.FILE_MISSING_BY_SLA), new[] { FileStatus.FILE_MISSING_BY_SLA } }
            };

        private const string None = "(none)";

        public static void AssertValidTransition(FileStatus? from, FileStatus to)
        {
            var key = from.HasValue ? from.Value.ToString() : None;
            FileStatus[] allowed;
            if (!ValidTransitions.TryGetValue(key, out allowed) || System.Array.IndexOf(allowed, to) < 0)
            {
                throw new InvalidStateTransitionException(from, to);
            }
        }
    }
}

File: apps/watcher/src/engine/state-transition.policy.ts
import type { FileStatus } from '@packages/contracts';
import { InvalidStateTransitionError } from '@packages/contracts';

const NONE = '(none)';

const VALID_TRANSITIONS: Record<string, FileStatus[]> = {
  [NONE]: ['FILE_DETECTED', 'FILE_MISSING_BY_SLA'],
  FILE_DETECTED: ['FILE_STABLE', 'FILE_STUCK'],
  FILE_STABLE: ['FILE_DUPLICATE'],
  FILE_STUCK: ['FILE_STABLE'],
  FILE_DUPLICATE: [],
  FILE_MISSING_BY_SLA: ['FILE_MISSING_BY_SLA'],
};

export function assertValidTransition(from: FileStatus | null, to: FileStatus): void {
  const key = from ?? NONE;
  const allowed = VALID_TRANSITIONS[key] ?? [];
  if (!allowed.includes(to)) {
    throw new InvalidStateTransitionError(from, to);
  }
}
